README.md
LC3 Tools
=========
This is a cleaned-up version of the LC-3 tools which has been modified slightly to install to `~/.lc3`, so it can be used alongside the [LCC-LC3 C Compiler](https://github.com/haplesshero13/lcc-lc3).

##Recent Changes
####May 2012

*Avery Yen*

Modified Makefile.def and configure to automatically install to `~/.lc3`. Also, `make install` now only installs executables. This is because having documents like a Makefile and a README in your executable path is a Bad Idea. The Makefile, README, and other documents previously installed by `make install` already live in the source directory. There doesn't seem to be any reason to make them live in two places.

Additionally, the lc3os and symbol tables now get installed to `~/.lc3/lc3os/`. This prevents them from being in the executable path, once again. lc3sim.c was modified to reflect this change.

## Description of Contents

The LC-3 tools package contains the lc3as assembler, the lc3sim simulator,
and lc3sim-tk, a Tcl/Tk-based GUI frontend to the simulator. All tools,
code, etc., were developed by Steven S. Lumetta on his own time with his
own resources for use in teaching classes based on the textbook by 
Yale N. Patt and Sanjay J. Patel entitled, *Introduction to Computing
Systems: From Bits & Gates to C & Beyond,* second edition, McGraw-Hill,
New York, New York, 2004, ISBN-0-07-246750-9.

The contents of the LC-3 tools distribution, including sources, management
tools, and data, are Copyright (c) 2003 Steven S. Lumetta.

The LC-3 tools distribution is free software covered by the GNU General 
Public License, and you are welcome to modify it and/or distribute copies 
of it under certain conditions. The file COPYING (distributed with the
tools) specifies those conditions. There is absolutely no warranty for 
the LC-3 tools distribution, as described in the file NO_WARRANTY (also
distributed with the tools).

## Necessary Tools

Installation requires versions of gcc (the Gnu C compiler),
flex (Gnu's version of lex), and wish (the Tcl/Tk graphical shell).
All of these tools are available for free from many sources.
If you have Gnu's readline installed, the configure script should
find it and use it for the command line version of the simulator.
I don't currently use the history feature, but will add it...someday.

Other necessary but more standard tools include uname, rm, cp, mkdir, 
and chmod.

Currently, the configure script searches only a few directories.
If your binaries are in a reasonable place that I overlooked, send
me a note and I'll add it to the default list. If you have 
idiosyncratic path names (e.g., the name of your fileserver in your
path), you will have to add the correct paths to the search path at 
the top of the configure script yourself.

N.B. I have installed the package on Cygwin, Solaris, and Linux
machines. Linux has been used by 2-3 other schools at the time of
the 0.5 release; Cygwin is stable on my home machine; Solaris GUI
version caused me grief last time I launched it, but I haven't
had time to investigate.


## Installation Instructions

###Windows
* Install Cygwin, making sure you have gcc, flex, and wish, as well as uname, rm, cp, mkdir, and chmod
* Now follow the general UNIX install

###Mac OS X 10.5+
* Install Xcode and the Xcode Developer Tools/Command Line Tools
* Check that you have gcc by typing `gcc --version`. If you get a message similar to `i686-apple-darwin11-llvm-gcc-4.2 (GCC) 4.2.1 (Based on Apple Inc. build 5658) (LLVM build 2336.9.00)` then you are all set.
* Now follow the general UNIX install

###Linux
* Install gcc, flex, and wish, using apt-get, yum, or whatever package manager your Linux system uses.
* Now follow the general UNIX install

###General UNIX
The LC-3 tools package is designed to work as either a personal or 
administrative installation on various flavors of Unix, including
Windows NT-based systems with appropriate support (e.g., Cygwin).

First, decide where the binaries and LC-3 OS code should be installed.

* If you want it in the ~/.lc3 directory, type `configure`.
* If you want it in a different directory, say `/usr/bin`, type `configure --installdir` `/usr/bin` replacing `/usr/bin` with the desired directory.

Then type `make install`.

## Maintainers
* Sean Smith (Dartmouth College)
* Avery Yen

## Original Author
* Steven S. Lumetta (lumetta@uiuc.edu).


CHANGE_LOG
5  May 2012
    Modified Makefile.def and configure to automatically install to `~/.lc3`.
    Additionally, the lc3os and symbol tables now get installed to `~/.lc3/lc3os/`.
    Also, `make install` now only installs executables.

12 February 2004 (version 0.9)
	fixed processing of hex input without leading "x" in lc3convert
	eliminated warnings for use of strlen as field width when
		disassembling instructions; strlen returns size_t on
		some systems
	cleared system breakpoint when simulator stops for any reason
		other than being interrupted by GUI, thus fixing bugs
		with breakpoint associated with the "next" command 
		surviving HALTs, user breakpoints, etc.

10 February 2004 (version 0.8)
	added missing quotes to MacOS-X Fink font overrides
	eliminated system breakpoints in favor of single address associated 
		with "next" command, thereby eliminating a couple of bugs: 
		(1) if the simulator was interrupted through interaction 
		with the GUI, such as setting/clearing breakpoints, during 
		a "next" operation, it lost track of where it was supposed 
		to stop; and
		(2) if the "clear all breakpoints" command was issued,
		the system breakpoint was erased (this bug was shadowed
		by the first)
	made the next and step commands correctly dump state after
		execution; delayed memory updates were not handled
		correctly for these commands
	made "Apply" button in options window pull font point size values
		from entry boxes before application of new options; the
		sizes were previously reset unless <Return> had been pressed
		in the entry box
	changed default code font for MacOS-X Fink installation to
		{LucidaTypewriter 11 normal}

9 February 2004 (version 0.7)
	set execution permission on lc3sim-tk during make
	allowed omission of "x" prefix when converting hexadecimal files
		with lc3convert
	fixed problems with CR-LF sequences in lc3as and lc3convert;
		both should now work (and were tested) with DOS-style
		end of line sequences
	added overrides for default fonts when installing for MacOS-X
		with Fink project's version of Tcl/Tk; new fonts
		suggested by Tevis Money, who also provided information
		on necessary paths, etc.

29 January 2004
	started to add support for MacOS, but found that two possible
		flavors exist (Fink and Aqua).  Fink needs some path
		additions and possibly default font changes, whereas
		Aqua apparently works with 0.6, and is available from 
		www.apple.com/downloads/macosx/unix_open_source/tcltkaqua.html

22 January 2004 (version 0.6)
	fixed a minor bug involving use of scan command functionality
		not available with Tk8.2 (but available with 8.3), thus
		preventing use with older versions of wish

18 December 2003 (version 0.5)
	(just comments by SSL today)
	I've heard no bug reports on Linux to date (2-3 installations?)
	I don't think that GUI works properly on Solaris, but I haven't
		had time to do a serious testing job at work yet.
	I use Cygwin at home, so it should be reasonably stable 
		(assuming that they don't insert bugs or expose any of mine)
	so...I guess I'm going to release more broadly, but I'll keep the
		0.5 as a warning for now
	(send bug reports to lumetta@uiuc.edu with something clear in the
		subject line)

12 November 2003
	added code to change BR to BRnzp in assembler
	allowed colons after labels, in the style used in some examples
		in the second edition of Patt and Patel; the colon form
		also allows the opcode and pseudo-op names to be used
		as labels (provided that they are followed by a colon
		when defined)

31 October 2003 (version 0.4)
	fixed off-by-one error in scanf limits in simulator
	fixed misunderstood argument in tcsetattr calls
	set VMIN and VTIME appropriately (1 and 0, resp.) when LC-3 runs
	finished writing and included new lc3convert tool for conversion 
		from .bin and .hex files to .obj files

30 October 2003
	started development of lc3convert tool by paring down copy of lc3as
		lexer

29 October 2003 (version 0.3)
	removed remaining dependencies on gmake (use of $@ and $<)
	added delay watch cursor use when opening options window
		(since it can take a long time when many fonts are
		available)
	added delay watch when starting up (takes a long time to
		fill 64k lines on some older machines)
	fixed ^C handling for Solaris
	fixed bugs with use of readline (in both Makefile and lc3sim.c)

21 October 2003
	added option to read LC-3 console input from stdin (or GUI
		console) during script execution; turned ON by default

19 October 2003 (version 0.2: fixed some bugs, large and small)
	re-disassembled code for GUI after writing memory
	created option to delay updated code to GUI until LC-3 stops
		(on by default)
	added support to GUI to change delay option
	fixed retagging PC and breakpoints when values change
	made simulator indicate breakpoints and presence of PC in
		disassembly; simulator also indicates breakpoints now
	added display of '\a', '\b', and a few others
	fixed display of '{' and '}' (Tcl can't parse braces as lists)
	removed symbols from region of memory into which new code is loaded
		in simulator
	disallowed duplicate labels in assembler
	KNOWN ISSUE: may want to translate \a, \b, etc. when sent to console
		(console doesn't act like a VT-100)

18 October 2003 (version 0.1)
	software packaged for release to Austin
	


lc3sim.h
/*									tab:8
 *
 * lc3sim.h - the main header file for the LC-3 simulator
 *
 * "Copyright (c) 2003 by Steven S. Lumetta."
 *
 * Permission to use, copy, modify, and distribute this software and its
 * documentation for any purpose, without fee, and without written 
 * agreement is hereby granted, provided that the above copyright notice
 * and the following two paragraphs appear in all copies of this software,
 * that the files COPYING and NO_WARRANTY are included verbatim with
 * any distribution, and that the contents of the file README are included
 * verbatim as part of a file named README with any distribution.
 * 
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE TO ANY PARTY FOR DIRECT, 
 * INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING OUT 
 * OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN IF THE AUTHOR 
 * HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 * 
 * THE AUTHOR SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING, BUT NOT 
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR 
 * A PARTICULAR PURPOSE.  THE SOFTWARE PROVIDED HEREUNDER IS ON AN "AS IS" 
 * BASIS, AND THE AUTHOR NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, 
 * UPDATES, ENHANCEMENTS, OR MODIFICATIONS."
 *
 * Author:	    Steve Lumetta
 * Version:	    1
 * Creation Date:   18 October 2003
 * Filename:	    lc3sim.h
 * History:
 *	SSL	1	18 October 2003
 *		Copyright notices and Gnu Public License marker added.
 */

#ifndef LC3SIM_H
#define LC3SIM_H

/* field access macros; "i" is an instruction */

#define F_DR(i)    (((i) >> 9) & 0x7)
#define F_SR1(i)   (((i) >> 6) & 0x7)
#define F_SR2(i)   (((i) >> 0) & 0x7)
#define F_BaseR(i) F_SR1
#define F_SR(i)    F_DR    /* for stores */
#define F_CC(i)    ((i) & 0x0E00)
#define F_vec8(i)  ((i) & 0xFF)

#define F_imm5(i)  (((i) & 0x010) == 0 ? ((i) & 0x00F) : ((i) | ~0x00F))
#define F_imm6(i)  (((i) & 0x020) == 0 ? ((i) & 0x01F) : ((i) | ~0x01F))
#define F_imm9(i)  (((i) & 0x100) == 0 ? ((i) & 0x1FF) : ((i) | ~0x1FF))
#define F_imm11(i) (((i) & 0x400) == 0 ? ((i) & 0x3FF) : ((i) | ~0x3FF))


/* instruction fields for output formatting */

typedef enum field_t field_t;
enum field_t {
    FMT_R1    = 0x001, /* DR or SR                      */
    FMT_R2    = 0x002, /* SR1 or BaseR                  */
    FMT_R3    = 0x004, /* SR2                           */
    FMT_CC    = 0x008, /* condition codes               */
    FMT_IMM5  = 0x010, /* imm5                          */
    FMT_IMM6  = 0x020, /* imm6                          */
    FMT_VEC8  = 0x040, /* vec8                          */
    FMT_ASC8  = 0x080, /* 8-bit ASCII                   */
    FMT_IMM9  = 0x100, /* label (or address from imm9)  */
    FMT_IMM11 = 0x200, /* label (or address from imm11) */
    FMT_IMM16 = 0x400  /* full instruction in hex       */
};


/* instruction formats for output */

typedef enum format_t format_t;
enum format_t {
    FMT_      = 0,
    FMT_RRR   = (FMT_R1 | FMT_R2 | FMT_R3),
    FMT_RRI   = (FMT_R1 | FMT_R2 | FMT_IMM5),
    FMT_CL    = (FMT_CC | FMT_IMM9),
    FMT_R     = FMT_R2,
    FMT_L     = FMT_IMM11,
    FMT_RL    = (FMT_R1 | FMT_IMM9),
    FMT_RRI6  = (FMT_R1 | FMT_R2 | FMT_IMM6),
    FMT_RR    = (FMT_R1 | FMT_R2),
    FMT_V     = FMT_VEC8,
    FMT_A     = FMT_ASC8,
    FMT_16    = FMT_IMM16
};


/* instruction flags */

typedef enum inst_flag_t inst_flag_t;
enum inst_flag_t {
    FLG_NONE = 0,
    FLG_SUBROUTINE = 1
};


/* LC-3 registers */

typedef enum reg_num_t reg_num_t;
enum reg_num_t {
	R_R0 = 0, R_R1, R_R2, R_R3, R_R4, R_R5, R_R6, R_R7,
	R_PC, R_IR, R_PSR,
	NUM_REGS
};

//extern int lc3_register[NUM_REGS];
#define REG(i) lc3_register[(i)]

extern int read_memory (int addr);
extern void write_memory (int addr, int value);


#endif /* LC3SIM_H */



lc3convert.f
/*									tab:8
 *
 * lc3convert.f - LC-3 binary and hexadecimal file conversion tool
 *
 * "Copyright (c) 2003 by Steven S. Lumetta."
 *
 * Permission to use, copy, modify, and distribute this software and its
 * documentation for any purpose, without fee, and without written 
 * agreement is hereby granted, provided that the above copyright notice
 * and the following two paragraphs appear in all copies of this software,
 * that the files COPYING and NO_WARRANTY are included verbatim with
 * any distribution, and that the contents of the file README are included
 * verbatim as part of a file named README with any distribution.
 * 
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE TO ANY PARTY FOR DIRECT, 
 * INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING OUT 
 * OF THE USE OF THIS SOFTWARE AND ITS DOCUMENTATION, EVEN IF THE AUTHOR 
 * HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 * 
 * THE AUTHOR SPECIFICALLY DISCLAIMS ANY WARRANTIES, INCLUDING, BUT NOT 
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR 
 * A PARTICULAR PURPOSE.  THE SOFTWARE PROVIDED HEREUNDER IS ON AN "AS IS" 
 * BASIS, AND THE AUTHOR NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT, 
 * UPDATES, ENHANCEMENTS, OR MODIFICATIONS."
 *
 * Author:	    Steve Lumetta
 * Version:	    2
 * Creation Date:   30 October 2003
 * Filename:	    lc3convert.f
 * History:
 *	SSL	2	31 October 2003
 *		Finished initial version.
 *	SSL	1	30 October 2003
 *		Started paring down lc3as code.
 */

%option noyywrap nounput

%{

#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>


static int line_num, num_errors, parse_hex, binary_value, bin_count;
static FILE* objout;

static void new_inst_line ();
static void bad_line ();
static void generate_bin_instruction (int value);
static void generate_hex_instruction (const char* val_str);

%}

/* 
   Operand types--note that hexadecimal does not require the "x" prefix
   for conversion.
*/
HEX      [xX]?[-]?[0-9a-fA-F]+
BINARY   [01]

/* comment and white space specification */
SPACE    [ \t]
COMMENT  [;][^\n\r]*
ENDLINE  {SPACE}*{COMMENT}?\r?\n\r?

/* exclusive lexing states to read binary and hexadecimal files */
%x ls_binary ls_hexadecimal ls_bin_garbage ls_hex_garbage

%%
    /* Choose appropriate parser. */
    if (parse_hex) {
        BEGIN (ls_hexadecimal);
    } else {
        BEGIN (ls_binary);
    }

    /* binary parser */
<ls_binary>{BINARY} {
    if (++bin_count == 17) {
        fprintf (stderr, "%3d: line contains more than 16 digits\n", line_num);
	num_errors++;
    } else {
        binary_value <<= 1;
	if (*yytext == '1')
	    binary_value++;
    }
}
<ls_binary>{ENDLINE} {
    if (bin_count == 0) { 
        /* a blank line */
	new_inst_line ();
    } else {
	if (bin_count < 16) {
	    fprintf (stderr, "%3d: line contains only %d digits\n", line_num,
		     bin_count);
	    num_errors++;
	}
	generate_bin_instruction (binary_value);
    }
}


    /* hexadecimal parser */
<ls_hexadecimal>{HEX}{ENDLINE} {generate_hex_instruction (yytext);}
<ls_hexadecimal>{ENDLINE} {new_inst_line (); /* a blank line */ }


    /* eat excess white space (both parsers) */
<ls_binary,ls_hexadecimal>{SPACE}+ {}  


    /* error handling (replicated because of substate use) */

<ls_binary>. {BEGIN (ls_bin_garbage);}
<ls_bin_garbage>[^\n\r]*{ENDLINE} {bad_line (); BEGIN (ls_binary);}

<ls_hexadecimal>. {BEGIN (ls_hex_garbage);}
<ls_hex_garbage>[^\n\r]*{ENDLINE} {bad_line (); BEGIN (ls_hexadecimal);}

%%

int
main (int argc, char** argv)
{
    int len, parse_error = 0;
    char* use_ext;
    char* ext;
    char* fname;

    if (argc == 3) {
	if (strcmp (argv[1], "-b2") == 0)
	    parse_hex = 0;
	else if (strcmp (argv[1], "-b16") == 0)
	    parse_hex = 1;
	else
	    parse_error = 1;
    } else
	parse_hex = 0;

    if (parse_error || argc < 2 || argc > 3) {
        fprintf (stderr, "usage: %s [-b2] <BIN filename>\n", argv[0]);
        fprintf (stderr, "       %s -b16 <HEX filename>\n", argv[0]);
	return 1;
    }

    /* Make our own copy of the filename. */
    len = strlen (argv[argc - 1]);
    if ((fname = malloc (len + 5)) == NULL) {
        perror ("malloc");
	return 3;
    }
    strcpy (fname, argv[argc - 1]);

    /* Check for .bin or .hex extension; if not found, add it. */
    use_ext = (parse_hex ? ".hex" : ".bin");
    if ((ext = strrchr (fname, '.')) == NULL || strcmp (ext, use_ext) != 0) {
	ext = fname + len;
        strcpy (ext, use_ext);
    }

    /* Open input file. */
    if ((lc3convertin = fopen (fname, "r")) == NULL) {
        fprintf (stderr, "Could not open %s for reading.\n", fname);
	return 2;
    }

    /* Open output files. */
    strcpy (ext, ".obj");
    if ((objout = fopen (fname, "w")) == NULL) {
        fprintf (stderr, "Could not open %s for writing.\n", fname);
	return 2;
    }

    line_num = 0;
    num_errors = 0;
    new_inst_line ();
    yylex ();
    printf ("%d errors found.\n", num_errors);
    if (num_errors > 0)
    	return 1;

    fclose (objout);

    return 0;
}

static void
new_inst_line () 
{
    binary_value = 0;
    bin_count = 0;
    line_num++;
}

static void 
bad_line ()
{
    fprintf (stderr, "%3d: contains unrecognizable characters\n",
	     line_num);
    num_errors++;
    new_inst_line ();
}

static int
read_val (const char* s, int* vptr, int bits)
{
    char* trash;
    long v;

    v = strtol (s, &trash, 16);
    if (v >= 0x8000)
        v |= 0xFFFF0000;
    if (v < -(1L << (bits - 1)) || v >= (1L << bits)) {
	fprintf (stderr, "%3d: constant outside of allowed range\n", line_num);
	num_errors++;
	return -1;
    }
    if ((v & (1UL << (bits - 1))) != 0)
	v |= ~((1UL << bits) - 1);
    *vptr = v;
    return 0;
}

static void
write_value (int val)
{
    unsigned char out[2];

    /* FIXME: just htons... */
    out[0] = (val >> 8);
    out[1] = (val & 0xFF);
    fwrite (out, 2, 1, objout);
}

static void 
generate_bin_instruction (int value)
{
    write_value (value);
    new_inst_line ();
}

static void 
generate_hex_instruction (const char* val_str)
{
    int value;

    if (!read_val (val_str, &value, 16))
        write_value (value);
    new_inst_line ();
}

